File: 백준/Gold/1707. 이분 그래프/이분 그래프.py
import sys

sys.setrecursionlimit(10**6)
input = sys.stdin.readline

N = int(input())
is_even = True


def dfs(node):
    global is_even
    visited[node] = True
    for i in A[node]:
        if not visited[i]:
            check[i] = (check[node] + 1) % 2
            dfs(i)
        elif check[node] == check[i]:
            is_even = False


for _ in range(N):
    v, e = map(int, input().split())
    A = [[] for _ in range(v + 1)]
    visited = [False] * (v + 1)
    check = [0] * (v + 1)
    is_even = True

    for i in range(e):
        s, e = map(int, input().split())
        A[s].append(e)
        A[e].append(s)

    for i in range(1, v + 1):
        if is_even:
            dfs(i)
        else:
            break

    if is_even:
        print("YES")
    else:
        print("NO")

File: 백준/Gold/1707. 이분 그래프/이분 그래프.js
let fs = require('fs');
const input = fs.readFileSync('./dev/stdin').toString().split('\n');

let t = Number(input[0]);
let line = 1;

const bfs = (x, graph, visited) => {
  let queue = [];
  queue.push(x);
  visited[x] = 0;
  while (queue.length > 0) {
    y = queue.pop();
    for (const z of graph[y]) {
      if (!visited[z]) {
        visited[z] = (visited[y] + 1) % 2;
        queue.push(z);
      }
    }
  }
};

const isBiapartite = (graph, visited) => {
  for (let x = 1; x < visited.length; x++) {
    for (const y of graph[x]) if (visited[x] === visited[y]) return false;
  }
  return true;
};

while (t--) {
  const [v, e] = input[line++].split(' ').map(Number);

  const graph = [];

  for (let i = 1; i <= v; i++) {
    graph[i] = [];
  }

  for (let i = 0; i < e; i++) {
    const [s, e] = input[line++].split(' ').map(Number);
    graph[s].push(e);
    graph[e].push(s);
  }

  const visited = Array(v + 1).fill(false);

  for (let i = 1; i <= v; i++) {
    if (!visited[i]) {
      bfs(i, graph, visited);
    }
  }

  if (isBiapartite(graph, visited)) console.log('YES');
  else console.log('NO');
}
